components/CharmCalculator.tsx

```diff
@@ -0,0 +1,198 @@
+"use client";
+
+import { useState } from "react";
+
+import { Select, SelectContent, SelectItem, SelectTrigger, SelectValue } from "@/components/ui/select";
+import { Table, TableBody, TableCaption, TableCell, TableHead, TableHeader, TableRow } from "@/components/ui/table";
+
+import Heroes from "../public/data/heroes.json";
+
+import { Charm, CharmPosition, StatSummary, GetStatSummary } from "@/lib/charm";
+
+export default function CharmCalculator() {
+  const availableClasses = Heroes.Classes.map((classType) => classType.Class);
+  const [selectedClass, setSelectedClass] = useState<String>(availableClasses[0]);
+
+  const [availableHeroes, setAvailableHeroes] = useState<any[] | undefined>(undefined);
+  const [selectedHero, setSelectedHero] = useState<String | undefined>(undefined);
+
+  const [availableCharms, setAvailableCharms] = useState<Charm[] | undefined>(undefined);
+  const [selectedLeftCharm, setSelectedLeftCharm] = useState<Charm | undefined>(undefined);
+  const [selectedMidCharm, setSelectedMidCharm] = useState<Charm | undefined>(undefined);
+  const [selectedRightCharm, setSelectedRightCharm] = useState<Charm | undefined>(undefined);
```

Comment:
i would model these as a single state variable, `selectedCharms` with an object.

```js
{ left: 'foo', mid: 'bar', right: 'baz' }
```

something along these lines.

then when you update the selected charm use a spread operator:

```js
setSelectedCharms({ ...selectedCharms, left: 'biz' })
```